File: polyglot/python/track_continuity_checker.py
"""
polyglot/python/track_continuity_checker.py

AIS Transponder-Gap Continuity Checker
======================================

Detects anomalies in AIS transponder reporting continuity. Identifies:
  - Time gaps (long pauses between reports)
  - Position jumps (unusual distance changes)
  - Velocity spikes (impossible speeds)

Scoring combines severity of each anomaly type into a single risk score.

Usage:
    from track_continuity_checker import ContinuityChecker
    
    checker = ContinuityChecker(
        time_gap_threshold=60,      # seconds before flagging
        pos_jump_threshold=500,     # meters for position jump
        velocity_max=45             # knots max reasonable speed
    )
    
    results = checker.check_tracks(tracks)
"""

import math
from dataclasses import dataclass, field
from datetime import datetime, timedelta
from typing import List, Optional, Tuple, Iterator, Dict, Any
from enum import Enum


class GapType(Enum):
    """Types of continuity anomalies detected."""
    TIME_GAP = "time_gap"           # Long pause in reporting
    POSITION_JUMP = "position_jump" # Unusual distance change
    VELOCITY_SPIKE = "velocity_spike"  # Impossible speed


@dataclass
class PositionReport:
    """Single AIS position report entry."""
    timestamp: datetime
    lat: float
    lon: float
    speed_over_ground: Optional[float] = None  # knots, optional
    course_over_ground: Optional[float] = None  # degrees, optional
    
    def to_tuple(self) -> Tuple[datetime, float, float]:
        return (self.timestamp, self.lat, self.lon)


@dataclass
class VesselTrack:
    """Container for a vessel's complete track history."""
    mmsi: str
    reports: List[PositionReport] = field(default_factory=list)
    
    def append(self, report: PositionReport):
        self.reports.append(report)
    
    @property
    def start_time(self) -> Optional[datetime]:
        return self.reports[0].timestamp if self.reports else None
    
    @property
    def end_time(self) -> Optional[datetime]:
        return self.reports[-1].timestamp if self.reports else None
    
    @property
    def duration_seconds(self) -> float:
        start = self.start_time
        end = self.end_time
        if start and end:
            return (end - start).total_seconds()
        return 0.0


@dataclass
class GapAnomaly:
    """Single detected anomaly event."""
    gap_type: GapType
    vessel_mmsi: str
    timestamp: datetime  # When the gap was observed
    severity_score: float  # 0-100, higher = worse
    
    def __str__(self) -> str:
        return f"{self.gap_type.value}: {self.severity_score:.1f}/100"


@dataclass
class TrackAnalysisResult:
    """Complete analysis result for one vessel."""
    mmsi: str
    total_reports: int
    duration_seconds: float
    avg_report_interval: float  # seconds between reports
    
    # Anomaly counts and scores
    time_gaps: List[GapAnomaly] = field(default_factory=list)
    position_jumps: List[GapAnomaly] = field(default_factory=list)
    velocity_spikes: List[GapAnomaly] = field(default_factory=list)
    
    # Aggregated metrics
    max_severity_score: float = 0.0
    total_anomalies: int = 0
    
    @property
    def overall_risk_score(self) -> float:
        """Composite risk score (0-100)."""
        if not self.time_gaps and not self.position_jumps and not self.velocity_spikes:
            return 0.0
        
        # Weighted average of max severities
        scores = [g.severity_score for g in self.time_gaps] + \
                 [g.severity_score for g in self.position_jumps] + \
                 [g.severity_score for g in self.velocity_spikes]
        
        if not scores:
            return 0.0
        
        # Normalize to 0-100 range, with more weight on recent issues
        max_possible = len(scores) * 100
        normalized = sum(scores) / max_possible * 100
        
        # Apply decay factor for older anomalies (less concerning)
        if self.duration_seconds > 0:
            age_factor = min(1.0, 3600 / (self.duration_seconds + 1))  # Max effect at 1 hour
            return normalized * age_factor
        
        return normalized


class ContinuityChecker:
    """
    Main engine for detecting AIS continuity anomalies.
    
    Configurable thresholds allow tuning for different operational contexts.
    """
    
    DEFAULT_CONFIG = {
        'time_gap_threshold': 60,      # seconds - flag if > this
        'position_jump_threshold': 500, # meters - flag if > this
        'velocity_max_knots': 45,       # max reasonable speed
        'velocity_spike_multiplier': 3.0,  # multiplier for spike detection
        'report_interval_normal': 120,   # expected normal interval (seconds)
        'report_interval_stddev': 60,    # acceptable variation
        'time_weight': 0.4,             # weight in severity calculation
        'pos_weight': 0.35,
        'vel_weight': 0.25,
    }
    
    def __init__(self, config: Optional[Dict[str, Any]] = None):
        self.config = {**self.DEFAULT_CONFIG, **(config or {})}
        
        # Pre-compute derived thresholds
        self.pos_threshold_meters = self.config['position_jump_threshold']
        self.vel_max_knots = self.config['velocity_max_knots']
        self.vel_spike_threshold = self.config['velocity_max_knots'] * \
                                  self.config['velocity_spike_multiplier']
        
    def check_tracks(self, tracks: List[VesselTrack]) -> List[TrackAnalysisResult]:
        """
        Analyze multiple vessel tracks and return results.
        
        Args:
            tracks: List of VesselTrack objects to analyze
            
        Returns:
            List of TrackAnalysisResult for each track
        """
        results = []
        
        for track in tracks:
            result = self._analyze_single_track(track)
            results.append(result)
        
        return results
    
    def check_stream(self, 
                     iterator: Iterator[PositionReport],
                     batch_size: int = 1000) -> Iterator[Tuple[VesselTrack, TrackAnalysisResult]]:
        """
        Process a stream of reports in batches for memory efficiency.
        
        Args:
            iterator: Stream of PositionReport objects
            batch_size: Number of reports to buffer before processing
            
        Yields:
            Tuple of (VesselTrack, TrackAnalysisResult) when batch is complete
        """
        current_track = None
        batch: List[PositionReport] = []
        
        for report in iterator:
            # Initialize track on first report
            if not current_track:
                current_track = VesselTrack(mmsi=report.lat, reports=[report])
            
            current_track.append(report)
            batch.append(report)
            
            if len(batch) >= batch_size:
                result = self._analyze_single_track(current_track)
                yield (current_track, result)
                batch.clear()
        
        # Handle remaining items in buffer
        if batch:
            current_track.reports.extend(batch)
            result = self._analyze_single_track(current_track)
            yield (current_track, result)
    
    def _analyze_single_track(self, track: VesselTrack) -> TrackAnalysisResult:
        """Analyze a single vessel's complete track."""
        if not track.reports:
            return TrackAnalysisResult(mmsi=track.mmsi, total_reports=0, 
                                       duration_seconds=0.0)
        
        # Initialize result with basic stats
        result = TrackAnalysisResult(
            mmsi=track.mmsi,
            total_reports=len(track.reports),
            duration_seconds=track.duration_seconds,
            avg_report_interval=track.duration_seconds / max(len(track.reports) - 1, 1)
        )
        
        # Analyze time gaps
        result.time_gaps = self._detect_time_gaps(track)
        
        # Analyze position jumps
        result.position_jumps = self._detect_position_jumps(track)
        
        # Analyze velocity spikes
        result.velocity_spikes = self._detect_velocity_spikes(track)
        
        # Calculate aggregated metrics
        result.max_severity_score = max(
            (g.severity_score for g in result.time_gaps), default=0.0,
            (g.severity_score for g in result.position_jumps), default=0.0,
            (g.severity_score for g in result.velocity_spikes), default=0.0,
        )
        
        result.total_anomalies = len(result.time_gaps) + \
                                len(result.position_jumps) + \
                                len(result.velocity_spikes)
        
        return result
    
    def _detect_time_gaps(self, track: VesselTrack) -> List[GapAnomaly]:
        """Detect gaps in reporting time sequence."""
        gaps = []
        prev_report = None
        
        for report in track.reports:
            if prev_report is None:
                prev_report = report
                continue
            
            delta = (report.timestamp - prev_report.timestamp).total_seconds()
            
            if delta > self.config['time_gap_threshold']:
                # Calculate severity based on how far over threshold
                excess_ratio = delta / self.config['time_gap_threshold']
                severity = min(100.0, 50 + (excess_ratio - 1) * 30)
                
                gaps.append(GapAnomaly(
                    gap_type=GapType.TIME_GAP,
                    vessel_mmsi=track.mmsi,
                    timestamp=report.timestamp,
                    severity_score=severity
                ))
            
            prev_report = report
        
        return gaps
    
    def _detect_position_jumps(self, track: VesselTrack) -> List[GapAnomaly]:
        """Detect unusual position changes between consecutive reports."""
        jumps = []
        
        for i in range(1, len(track.reports)):
            prev = track.reports[i - 1]
            curr = track.reports[i]
            
            # Calculate distance (Haversine approximation)
            delta_lat = math.radians(curr.lat - prev.lat)
            delta_lon = math.radians(curr.lon - prev.lon)
            
            a = (math.sin(delta_lat/2)**2 + 
                 math.cos(math.radians(prev.lat)) * math.cos(math.radians(curr.lat)) *
                 math.sin(delta_lon/2)**2)
            
            c = 2 * math.asin(math.sqrt(a))
            distance_meters = 6371000 * c
            
            if distance_meters > self.pos_threshold_meters:
                # Severity based on jump size relative to threshold
                ratio = distance_meters / self.pos_threshold_meters
                severity = min(100.0, 40 + (ratio - 1) * 25)
                
                jumps.append(GapAnomaly(
                    gap_type=GapType.POSITION_JUMP,
                    vessel_mmsi=track.mmsi,
                    timestamp=curr.timestamp,
                    severity_score=severity
                ))
        
        return jumps
    
    def _detect_velocity_spikes(self, track: VesselTrack) -> List[GapAnomaly]:
        """Detect impossible or unusual velocity changes."""
        spikes = []
        
        for i in range(1, len(track.reports)):
            prev = track.reports[i - 1]
            curr = track.reports[i]
            
            # Calculate time delta (avoid division by zero)
            if prev.timestamp == curr.timestamp:
                continue
            
            time_delta_hours = (curr.timestamp - prev.timestamp).total_seconds() / 3600.0
            
            # Calculate distance
            delta_lat = math.radians(curr.lat - prev.lat)
            delta_lon = math.radians(curr.lon - prev.lon)
            
            a = (math.sin(delta_lat/2)**2 + 
                 math.cos(math.radians(prev.lat)) * math.cos(math.radians(curr.lat)) *
                 math.sin(delta_lon/2)**2)
            
            c = 2 * math.asin(math.sqrt(a))
            distance_miles = 3440.5 * c
            
            # Calculate speed in knots (nautical miles per hour)
            if time_delta_hours > 0:
                speed_knots = distance_miles / time_delta_hours
                
                if speed_knots > self.vel_spike_threshold:
                    ratio = speed_knots / self.vel_spike_threshold
                    severity = min(100.0, 30 + (ratio - 1) * 25)
                    
                    spikes.append(GapAnomaly(
                        gap_type=GapType.VELOCITY_SPIKE,
                        vessel_mmsi=track.mmsi,
                        timestamp=curr.timestamp,
                        severity_score=severity
                    ))
        
        return spikes
    
    def filter_by_min_severity(self, results: List[TrackAnalysisResult], 
                               min_score: float = 20.0) -> List[TrackAnomaly]:
        """
        Filter anomalies by minimum severity score.
        
        Args:
            results: Analysis results to filter
            min_score: Minimum severity threshold (0-100)
            
        Returns:
            Flattened list of GapAnomaly objects above threshold
        """
        all_anomalies = []
        
        for result in results:
            if not result.time_gaps and not result.position_jumps and \
               not result.velocity_spikes:
                continue
            
            for gap in result.time_gaps + result.position_jumps + result.velocity_spikes:
                if gap.severity_score >= min_score:
                    all_anomalies.append(gap)
        
        return sorted(all_anomalies, key=lambda x: -x.severity_score)


# =============================================================================
# Demo / Self-Contained Test Data Generator and Runner
# =============================================================================

def generate_demo_tracks(num_vessels: int = 5, 
                         num_reports_per_vessel: int = 100,
                         include_anomalies: bool = True) -> List[VesselTrack]:
    """Generate realistic demo track data with optional anomalies."""
    import random
    
    tracks = []
    
    for vessel_idx in range(num_vessels):
        # Create a base trajectory (simple linear path with noise)
        start_lat, start_lon = 35.0 + vessel_idx * 0.1, -120.0 + vessel_idx * 0.1
        end_lat, end_lon = 36.0, -121.0
        
        # Base speed in knots
        base_speed_knots = 15 + random.uniform(-3, 3)
        
        track = VesselTrack(mmsi=f"367{vessel_idx:04d}")
        
        current_time = datetime.now() - timedelta(hours=num_reports_per_vessel)
        
        for report_idx in range(num_reports_per_vessel):
            # Add some randomness to reporting intervals (normal 120s +/- 30s)
            if report_idx == 0:
                interval = random.uniform(90, 150)
            else:
                expected_time = current_time + timedelta(seconds=report_idx * 120)
                
                # Occasionally create a time gap (simulates transponder off)
                if include_anomalies and random.random() < 0.15:
                    gap_duration = random.uniform(180, 900)  # 3-15 minutes
                    expected_time += timedelta(seconds=gap_duration)
                
                interval = (expected_time - current_time).total_seconds()
            
            # Add small position noise (simulates GPS error)
            lat_noise = random.gauss(0.0, 0.02)  # ~2km max
            lon_noise = random.gauss(0.0, 0.02)
            
            report_lat = start_lat + (end_lat - start_lat) * (report_idx / num_reports_per_vessel) + lat_noise
            report_lon = start_lon + (end_lon - start_lon) * (report_idx / num_reports_per_vessel) + lon_noise
            
            # Occasionally create a position jump (simulates satellite handoff error)
            if include_anomalies and random.random() < 0.05:
                jump_lat = random.gauss(0.0, 0.15)  # ~15km max
                jump_lon = random.gauss(0.0, 0.15)
                report_lat += jump_lat
                report_lon += jump_lon
            
            report = PositionReport(
                timestamp=current_time + timedelta(seconds=interval),
                lat=report_lat,
                lon=report_lon,
                speed_over_ground=base_speed_knots + random.gauss(0, 1)
            )
            
            track.append(report)
            current_time = expected_time
        
        tracks.append(track)
    
    return tracks


def main():
    """Run self-contained demo with realistic data."""
    print("=" * 60)
    print("AIS Track Continuity Checker - Demo")
    print("=" * 60)
    
    # Generate and analyze demo data
    tracks = generate_demo_tracks(num_vessels=10, num_reports_per_vessel=200)
    
    # Configure

File: polyglot/typescript/track_continuity_checker.ts
import { AISTrackPoint, TrackContinuityResult, ContinuityConfig } from './types';

export const DEFAULT_CONFIG: ContinuityConfig = {
  maxTimeGapSeconds: 60 * 5, // 5 minutes between points before flagging as gap
  maxSpeedChangeKnots: 20,    // Max reasonable speed change per interval
  minSOGForValidPoint: 0.1,   // Minimum SOG to consider point valid (avoid static noise)
  positionJumpThresholdNauticalMiles: 5, // Max jump before flagging as anomaly
};

export function createContinuityChecker(config?: ContinuityConfig): TrackContinuityChecker {
  const cfg = { ...DEFAULT_CONFIG, ...(config || {}) };
  
  return {
    checkTrack,
    analyzeGap,
    calculateAnomalyScore,
  };
}

function createResult(
  points: AISTrackPoint[],
  gaps: GapInfo[],
  anomalies: AnomalyInfo[],
  summary: TrackSummary
): TrackContinuityResult {
  return {
    points,
    gaps,
    anomalies,
    summary,
    metadata: {
      checkedAt: new Date().toISOString(),
      configUsed: cfg,
      totalPoints: points.length,
    },
  };
}

function analyzeGap(
  prevPoint: AISTrackPoint | null,
  currPoint: AISTrackPoint,
  index: number
): GapInfo | undefined {
  if (!prevPoint) return undefined;
  
  const timeDiffSeconds = (currPoint.timestamp.getTime() - prevPoint.timestamp.getTime()) / 1000;
  
  // Check if gap exceeds threshold
  if (timeDiffSeconds > cfg.maxTimeGapSeconds) {
    return {
      index,
      prevIndex: index - 1,
      timeDeltaSeconds: timeDiffSeconds,
      isCritical: timeDiffSeconds > cfg.maxTimeGapSeconds * 2,
      estimatedDistanceTraveled: calculateEstimatedTravel(
        prevPoint,
        currPoint,
        timeDiffSeconds
      ),
    };
  }
  
  return undefined;
}

function analyzeSpeedChange(prevPoint: AISTrackPoint, currPoint: AISTrackPoint): AnomalyInfo | undefined {
  if (!prevPoint || !currPoint) return undefined;
  
  const speedDiff = Math.abs(currPoint.sog - prevPoint.sog);
  
  // Flag extreme speed changes
  if (speedDiff > cfg.maxSpeedChangeKnots) {
    return {
      type: 'SPEED_CHANGE',
      index: currPoint.index,
      severity: speedDiff / cfg.maxSpeedChangeKnots,
      details: {
        prevSpeed: prevPoint.sog,
        currentSpeed: currPoint.sog,
        change: speedDiff,
      },
    };
  }
  
  return undefined;
}

function analyzePositionJump(prevPoint: AISTrackPoint, currPoint: AISTrackPoint): AnomalyInfo | undefined {
  if (!prevPoint || !currPoint) return undefined;
  
  const distance = calculateDistance(prevPoint.lat, prevPoint.lon, currPoint.lat, currPoint.lon);
  
  // Flag impossible position jumps
  if (distance > cfg.positionJumpThresholdNauticalMiles) {
    const timeDiffHours = (currPoint.timestamp.getTime() - prevPoint.timestamp.getTime()) / (1000 * 3600);
    
    return {
      type: 'POSITION_JUMP',
      index: currPoint.index,
      severity: distance / cfg.positionJumpThresholdNauticalMiles,
      details: {
        previousPosition: `${prevPoint.lat.toFixed(4)}, ${prevPoint.lon.toFixed(4)}`,
        currentPosition: `${currPoint.lat.toFixed(4)}, ${currPoint.lon.toFixed(4)}`,
        distanceNauticalMiles: distance,
        timeDeltaHours: timeDiffHours,
        impliedSpeedKnots: timeDiffHours > 0 ? (distance / timeDiffHours).toFixed(2) : 'N/A',
      },
    };
  }
  
  return undefined;
}

function analyzeStaticNoise(point: AISTrackPoint): AnomalyInfo | undefined {
  if (!point.sog || point.sog < cfg.minSOGForValidPoint) {
    // Very low or zero SOG might indicate static noise or sensor issue
    if (point.sog === 0 || (point.sog && point.sog < 0.1)) {
      return {
        type: 'STATIC_NOISE',
        index: point.index,
        severity: 1 - (point.sog / cfg.minSOGForValidPoint), // Higher severity for lower SOG
        details: {
          sog: point.sog,
          note: 'Near-zero speed detected',
        },
      };
    }
  }
  
  return undefined;
}

function calculateEstimatedTravel(
  prevPoint: AISTrackPoint,
  currPoint: AISTrackPoint,
  timeDiffSeconds: number
): { distanceNauticalMiles: number; avgSpeedKnots: number } | null {
  if (timeDiffSeconds <= 0) return null;
  
  const distance = calculateDistance(prevPoint.lat, prevPoint.lon, currPoint.lat, currPoint.lon);
  const avgSpeed = timeDiffSeconds > 0 ? (distance / (timeDiffSeconds / 3600)) : 0;
  
  return {
    distanceNauticalMiles: distance,
    avgSpeedKnots: avgSpeed,
  };
}

function calculateDistance(lat1: number, lon1: number, lat2: number, lon2: number): number {
  const R = 3440.065; // Earth radius in nautical miles
  
  const dLat = (lat2 - lat1) * Math.PI / 180;
  const dLon = (lon2 - lon1) * Math.PI / 180;
  
  const a = 
    Math.sin(dLat/2) * Math.sin(dLat/2) +
    Math.cos(lat1 * Math.PI / 180) * Math.cos(lat2 * Math.PI / 180) *
    Math.sin(dLon/2) * Math.sin(dLon/2);
  
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  
  return R * c;
}

function calculateTrackSummary(points: AISTrackPoint[]): TrackSummary {
  if (points.length === 0) {
    return {
      totalPoints: 0,
      timeSpanHours: 0,
      avgSpeedKnots: 0,
      maxSpeedKnots: 0,
      minSpeedKnots: 0,
      distanceCoveredNauticalMiles: 0,
    };
  }
  
  const sorted = [...points].sort((a, b) => 
    a.timestamp.getTime() - b.timestamp.getTime()
  );
  
  const timeSpanHours = (sorted[sorted.length - 1].timestamp.getTime() - sorted[0].timestamp.getTime()) / (1000 * 3600);
  const avgSpeed = sorted.reduce((sum, p) => sum + p.sog || 0, 0) / sorted.length;
  
  return {
    totalPoints: points.length,
    timeSpanHours: Math.max(0, timeSpanHours),
    avgSpeedKnots: avgSpeed,
    maxSpeedKnots: Math.max(...sorted.map(p => p.sog || 0)),
    minSpeedKnots: Math.min(...sorted.map(p => p.sog || Infinity)),
    distanceCoveredNauticalMiles: calculateDistance(sorted[0].lat, sorted[0].lon, 
      sorted[sorted.length - 1].lat, sorted[sorted.length - 1].lon),
  };
}

function calculateAnomalyScore(
  result: TrackContinuityResult,
  baseWeight: number = 50
): number {
  let score = baseWeight;
  
  // Gap penalties (heavier for critical gaps)
  const gapPenalty = result.gaps.reduce((sum, g) => 
    sum + (g.isCritical ? 15 : 8), 0);
  score -= gapPenalty * 2;
  
  // Speed change penalties
  const speedChangePenalty = result.anomalies.filter(a => a.type === 'SPEED_CHANGE')
    .reduce((sum, a) => sum + (a.severity - 1) * 5, 0);
  score -= speedChangePenalty;
  
  // Position jump penalties (very heavy as they indicate major issues)
  const jumpPenalty = result.anomalies.filter(a => a.type === 'POSITION_JUMP')
    .reduce((sum, a) => sum + (a.severity - 1) * 20, 0);
  score -= jumpPenalty;
  
  // Static noise penalties
  const staticPenalty = result.anomalies.filter(a => a.type === 'STATIC_NOISE')
    .reduce((sum, a) => sum + (a.severity - 1) * 3, 0);
  score -= staticPenalty;
  
  return Math.max(0, Math.min(100, score)); // Clamp to 0-100
}

export interface GapInfo {
  index: number;
  prevIndex: number;
  timeDeltaSeconds: number;
  isCritical: boolean;
  estimatedDistanceTraveled?: number;
}

export interface AnomalyInfo {
  type: 'SPEED_CHANGE' | 'POSITION_JUMP' | 'STATIC_NOISE';
  index: number;
  severity: number; // 0-1 scale, higher = worse
  details: Record<string, any>;
}

export interface TrackSummary {
  totalPoints: number;
  timeSpanHours: number;
  avgSpeedKnots: number;
  maxSpeedKnots: number;
  minSpeedKnots: number;
  distanceCoveredNauticalMiles: number;
}

export interface TrackContinuityResult {
  points: AISTrackPoint[];
  gaps: GapInfo[];
  anomalies: AnomalyInfo[];
  summary: TrackSummary;
  metadata: {
    checkedAt: string;
    configUsed: ContinuityConfig;
    totalPoints: number;
  };
}

export interface ContinuityConfig {
  maxTimeGapSeconds?: number;
  maxSpeedChangeKnots?: number;
  minSOGForValidPoint?: number;
  positionJumpThresholdNauticalMiles?: number;
}

interface TrackContinuityChecker {
  checkTrack: (points: AISTrackPoint[], config?: ContinuityConfig) => TrackContinuityResult;
  analyzeGap: (prev: AISTrackPoint | null, curr: AISTrackPoint, index: number) => GapInfo | undefined;
  calculateAnomalyScore: (result: TrackContinuityResult, baseWeight?: number) => number;
}

// ============================================================================
// ENTRY POINT / DEMO
// ============================================================================

function createDemoData(): AISTrackPoint[] {
  const now = new Date();
  
  // Create a realistic vessel track with some anomalies for testing
  
  const baseLat = 35.0;
  const baseLon = -125.0;
  
  let lat = baseLat;
  let lon = baseLon;
  let sog = 15.0;
  let cog = 90.0; // Eastward
  
  const points: AISTrackPoint[] = [];
  
  for (let i = 0; i < 24; i++) {
    // Simulate movement with some noise and anomalies
    lat += Math.sin(i * 0.5) * 0.01 + 0.002;
    lon += Math.cos(i * 0.5) * 0.01 + 0.003;
    
    // Add realistic SOG variation
    sog = 14.0 + (Math.random() - 0.5) * 2.0;
    
    // Inject anomalies at specific points
    
    // Anomaly 1: Gap at point 8-9 (about 7 minutes gap)
    if (i === 8) {
      sog = 16.0;
    } else if (i === 9) {
      // Jump forward in time to simulate a 7-minute gap
      const baseTime = new Date(now.getTime() - (24 * i) * 60000);
      points.push({
        index: i,
        timestamp: new Date(baseTime.getTime() + 7 * 60000), // 7 minute jump!
        lat: parseFloat((lat).toFixed(5)),
        lon: parseFloat((lon).toFixed(5)),
        sog: sog,
        cog: 90.0,
      });
    } else {
      points.push({
        index: i,
        timestamp: new Date(now.getTime() - (24 * i) * 60000),
        lat: parseFloat((lat).toFixed(5)),
        lon: parseFloat((lon).toFixed(5)),
        sog: sog,
        cog: 90.0,
      });
    }
    
    // Anomaly 2: Speed spike at point 15
    if (i === 15) {
      sog = 35.0; // Unusual speed increase
    }
    
    // Anomaly 3: Position jump at point 20
    if (i === 20) {
      lat += 0.5; // Jump half a degree south!
    }
    
    // Anomaly 4: Static noise at point 22
    if (i === 22) {
      sog = 0.1; // Near-zero speed
    }
  }
  
  return points;
}

function runDemo(): void {
  console.log('='.repeat(60));
  console.log('AIS Track Continuity Checker - Demo');
  console.log('='.repeat(60));
  console.log();
  
  // Create demo data with known anomalies
  const demoPoints = createDemoData();
  
  // Check continuity with default config
  const checker = createContinuityChecker();
  const result = checker.checkTrack(demoPoints);
  
  // Calculate anomaly score
  const anomalyScore = checker.calculateAnomalyScore(result, 50);
  
  console.log('TRACK SUMMARY');
  console.log('-'.repeat(40));
  console.log(`Total Points:    ${result.summary.totalPoints}`);
  console.log(`Time Span:       ${(result.summary.timeSpanHours).toFixed(2)} hours`);
  console.log(`Avg Speed:       ${(result.summary.avgSpeedKnots).toFixed(1)} knots`);
  console.log(`Max Speed:       ${(result.summary.maxSpeedKnots).toFixed(1)} knots`);
  console.log(`Min Speed:       ${(result.summary.minSpeedKnots).toFixed(1)} knots`);
  console.log(`Distance:        ${(result.summary.distanceCoveredNauticalMiles).toFixed(2)} NM`);
  console.log();
  
  console.log('GAP ANALYSIS');
  console.log('-'.repeat(40));
  if (result.gaps.length === 0) {
    console.log('No significant gaps detected.');
  } else {
    result.gaps.forEach((gap, idx) => {
      const status = gap.isCritical ? '⚠️ CRITICAL' : '📍';
      console.log(`${idx + 1}. ${status} - Gap of ${(gap.timeDeltaSeconds).toFixed(0)} seconds`);
      if (gap.estimatedDistanceTraveled) {
        console.log(`   Estimated travel: ${gap.estimatedDistanceTraveled.toFixed(2)} NM @ ${(gap.estimatedDistanceTraveled / (gap.timeDeltaSeconds/3600)).toFixed(1)} knots`);
      }
    });
  }
  console.log();
  
  console.log('ANOMALY DETECTION');
  console.log('-'.repeat(40));
  if (result.anomalies.length === 0) {
    console.log('No anomalies detected.');
  } else {
    result.anomalies.forEach((anomaly, idx) => {
      const typeLabel = anomaly.type;
      const severityPct = (anomaly.severity * 100).toFixed(1);
      
      let detailText = '';
      switch (anomaly.type) {
        case 'SPEED_CHANGE':
          detailText = `Speed changed from ${anomaly.details.prevSpeed.toFixed(1)} to ${anomaly.details.currentSpeed.toFixed(1)} knots (${anomaly.details.change.toFixed(1)} knot change)`;
          break;
        case 'POSITION_JUMP':